Sources/AsyncHTTPClient/ConnectionPool/HTTPConnectionPool+Factory.swift

```diff
@@ -0,0 +1,314 @@
+                    channel.pipeline.removeHandler(decoder).flatMap {
+                        channel.pipeline.removeHandler(encoder)
+                    }
+                }
+            }.flatMap { () -> EventLoopFuture<(Channel, HTTPVersion)> in
+                switch self.key.scheme {
+                case .unix, .http_unix, .https_unix:
+                    preconditionFailure("Unexpected scheme. Not supported for proxy!")
+                case .http:
+                    return channel.eventLoop.makeSucceededFuture((channel, .http1_1))
+                case .https:
+                    var tlsConfig = self.tlsConfiguration
+                    // since we can support h2, we need to advertise this in alpn
+                    tlsConfig.applicationProtocols = ["http/1.1" /* , "h2" */ ]
+                    let tlsEventHandler = TLSEventsHandler()
+
+                    let sslContextFuture = self.sslContextCache.sslContext(
+                        tlsConfiguration: tlsConfig,
+                        eventLoop: channel.eventLoop,
+                        logger: logger
+                    )
+
+                    return sslContextFuture.flatMap { sslContext -> EventLoopFuture<String?> in
+                        do {
+                            let sslHandler = try NIOSSLClientHandler(
+                                context: sslContext,
+                                serverHostname: self.key.host
+                            )
+                            try channel.pipeline.syncOperations.addHandler(sslHandler)
+                            try channel.pipeline.syncOperations.addHandler(tlsEventHandler)
+                            return tlsEventHandler.tlsEstablishedFuture
+                        } catch {
+                            return channel.eventLoop.makeFailedFuture(error)
+                        }
+                    }.flatMap { negotiated -> EventLoopFuture<(Channel, HTTPVersion)> in
+                        channel.pipeline.removeHandler(tlsEventHandler).map {
+                            switch negotiated {
+                            case "h2":
+                                return (channel, .http2)
+                            default:
+                                return (channel, .http1_1)
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private func makePlainBootstrap(eventLoop: EventLoop) -> NIOClientTCPBootstrapProtocol {
+        #if canImport(Network)
+            if #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *), let tsBootstrap = NIOTSConnectionBootstrap(validatingGroup: eventLoop) {
+                return tsBootstrap
+                    .addTimeoutIfNeeded(self.clientConfiguration.timeout)
+                    .channelInitializer { channel in
+                        do {
+                            try channel.pipeline.syncOperations.addHandler(HTTPClient.NWErrorHandler())
```

Comment:
wouldn't async operations look cleaner here?

Reply:
It would look nicer, I agree. 

However with the current impl. we don't need to alloc an `EventLoopFuture` at all, since we can return the prealloced `return channel.eventLoop.makeSucceededFuture(())`.

This would not necessarily be the case when adding the handlers async.

Reply:
nit: then we should probably return `eventLoop.makeSucceededVoidFuture()`